### ISSUE TITLE
RoundRobin in recovery is not reset when main branch success

### ISSUE THREAD

**Issue by FrGrQuim** [ros-navigation/navigation2#4737]
## Bug report

**Required Info:**

- Operating System:
  - Custom Linux
- ROS2 Version:
  - Iron source
- Version or commit hash:
  - 1.2.9
- DDS implementation:
  - CycloneDDS

#### Steps to reproduce issue
In our Behavior Tree, we have a RecoveryNode with a RoundRobin in the recovery branch, structured as follows:

![revovery_round_robin drawio](https://github.com/user-attachments/assets/d39274e5-d995-4258-b8b5-edb2e5245923)

#### Expected behavior
We expect that each time the main action succeeds, the round_robin is reset. Therefore, the next time the main action fails, the first recovery action is triggered.

This behavior is illustrated by the following sequence:
1 - Main action: Fail
2 - Recovery A: Success
3 - Main action: Fail
4 - Recovery B: Success
5 - Main action: Success
... (later)
6 - Main action: Fail
7 - **Recovery A is call**
#### Actual behavior
Currently, we observe that the roundRobin is not reset when the main branch succeeds. Therefore, the next time the main action fails, the roundRobin continues incrementing the child index.

This behavior is illustrated by the following sequence:
1 - Main action: Fail
2 - Recovery A: Success
3 - Main action: Fail
4 - Recovery B: Success
5 - Main action: Success
... (later)
6 - Main action: Fail
7 - **Recovery C is call**

#### Additional information
After checking the code, we noticed that the only way to reset the roundRobin is by calling the halt function. However, in the BehaviorTreeCPP library, the halt function is only called when the target node is in the RUNNING state (which seems consistent with the function name). The issue is that the Recovery node does not call the roundRobin's halt function when it is in the RUNNING state.

My question, then, is: Is this the intended behavior? And if so, do you have any suggestions on how to implement the expected behavior sequence?

**Comment by SteveMacenski** [comment]
This seems to be the intended behavior as also evidenced by the documentation: https://docs.nav2.org/behavior_trees/overview/nav2_specific_nodes.html#control-roundrobin, but this is worth some discussion and potential behavior changes (I'd suggest reading this page before continuing, I myself refer to it in these questions). The text below is a bit of my own logical process to document how I came to my conclusions / suggestions for potential improvements.

The control node in a BT regulates the execution of the child nodes it has access to (in this case, your RoundRobin control node has child nodes of Recoveries A, B, and C & your RecoveryNode has children of your main action and the RoundRobin).  

The result of the main action is not communicated to the RoundRobin directly as they are both children of the same parent node and are thus independent of each other to communicate statuses. Rather the handling of RoundRobin's state based on the state of the main action is on the logic of the parent they share, i.e. `RecoveryNode`. 

It sounds like you would like the `RecoveryNode` to reset the state of the recovery child (second child, in this case `RoundRobin`) when the primary task (in your case, `main action`) succeeds. When [`halt()` is called in the RoundRobin](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/round_robin_node.cpp#L85-L90), we seem to already reset the state of the child IDs to start back at its first child node (RecoveryA) like you want.  

So it seems that the behavior you want is really tied into the fact that `halt()` isn't being called on the recovery child of the `RecoveryNode` once the primary task succeeds. In implementation, `ControlNode::haltChild(1);`. Right now, we call [RecoveryNode's halt()](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L119-L124) which resets its state and calls `ControlNode::halt()`, but that only calls `halt()` on currently `RUNNING` nodes. There's been discussions about that in BT.CPP, but ultimately that's the right move according to the maintainers. 

So, I think perhaps adding `ControlNode::haltChild(1);` when [the first child is successful](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L53-L56) would fix that problem. I would appreciate if you tested that to verify this resolves the behavior issue you expect and get back to me.

The follow up question is: should we make this change in the `RecoveryNode` in Nav2 by default. We shouldn't halt the recovery child each time we exit the RecoveryNode (since we want to maintain some state on `FAILURE` to tick the next node), but would `SUCCESS` be a special case given the intent of the BT Node's second child as a recovery mechanism? Would we also have some `SUCCESS` state being maintained - perhaps if we're doing something like a `PipelineSequence` or other Control Node as a container for more complex logic under the RecoveryNode's second child (i.e. recovery child). 

I think ... possibly so, someone might want to maintain some state. But also, if the action is now successful, wouldn't it make sense to start again from the top - given the explicit intent of the `RecoveryNode`? So I'm leaning towards approving that change for Nav2's RecoveryNode more generally. I think restarting the recovery child's work each time its newly failing is sensible for the design and intent.

Let me know what you think!

**Comment by SteveMacenski** [comment]
@tonynajjar what do you think? I know you're thinking about BTs these days

**Comment by SteveMacenski** [comment]
@tonynajjar mind giving this your 2 cents?

**Comment by tonynajjar** [comment]
I see it the same that it would make sense for the RecoveryNode should halt the second child (recovery action). But maybe it should be done at both FAILURE and SUCCESS of the second child.

> We shouldn't halt the recovery child each time we exit the RecoveryNode (since we want to maintain some state on FAILURE to tick the next node)

Not sure I get what you mean here.

Let's take this example:

```
  <RecoveryNode number_of_retries="2">
      <MainAction/>
      <RetryUntilSuccessful num_attempts="5">
        <RecoveryAction/>
      </RetryUntilSuccessful>
  </RecoveryNode>
```

From intuition, in this example I would expect the RecoveryAction to always be tried up to 5 times (i.e. RetryUntilSuccessful is halted/reseted at every terminal state). The scenarios are:

 - MainAction fails, RecoveryAction succeeds after n attempts, MainAction fails again -> RetryUntilSuccessful should have 5 attempts remaining 

 - MainAction fails, RecoveryAction succeeds after n attempts, MainAction succeeds, RecoveryNode succeeds -> then when we tick the whole thing again, RetryUntilSuccessful should have 5 attempts remaining 

Maybe there are other scenarios where you would intuitively want the opposite? Or maybe we don't even share the same intuition on this example?

**Comment by SteveMacenski** [comment]
I'm suggesting that if the first child (main action) is SUCCESSFUL, we halt the second child (recovery action) via `ControlNode::haltChild(1);` [here in the code](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L53-L56) since the recovery is meant to make the first child (main action) pass again, so its completed its job and can be reset.  This is not currently happening and I think this is what the issue is mentioning. 

Do we have agreement on that?

> I see it the same that it would make sense for the RecoveryNode should halt the second child (recovery action). But maybe it should be done at both FAILURE and SUCCESS of the second child.

This is a bit different from what I think I/the author was thinking. Right now, [we do that in SUCCESS of the recovery action](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L97), but not failure. 

> MainAction fails, RecoveryAction succeeds after n attempts, MainAction succeeds, RecoveryNode succeeds -> then when we tick the whole thing again, RetryUntilSuccessful should have 5 attempts remaining

Agreed 👍 I think that's the modification I mention in the first paragraph that I think we're on the same page about making. 

> MainAction fails, RecoveryAction succeeds after n attempts, MainAction fails again -> RetryUntilSuccessful should have 5 attempts remaining

Here I think disagree. The RecoveryAction succeeding means that it successfully performed a recovery action, not that this recovery action actually successfully got the MainAction to work again. If multiple MainAction failures happen in sequence, I don't think the RetryUntilSuccessful/RecoveryAction should be reset/halted. The number of times in a row to try to execute the recovery should be capped at 5x total between successes to reset the counter for a next recovery 'session' of the MainAction. 

But, I'm open to push back here. I just think the recovery action's job is to make the main action succeed. If it hasn't done that yet, its state should be maintained so that things like PipelineSequence / Round Robin can do their thing and maintain their appropriate states. 

That obviously changes once we fail MainAction again, because the 2x retries in the RecoveryNode [would reset the first child](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L63-L66). **But, shouldn't we actually reset both main action and recovery action children so that a retry gets a clean, new state retry?**

---

In summary, I think we should:
* Halt the recovery child (1) on main action (0)'s success. Today, nothing is done. 
* Halt both children (0, 1) on retry. Today, only (0) is halted.

That would reset the recovery state to fresh once it successfully makes the main action succeed and its done its job. Also, if a retry is made, that retry has a fresh state in both the main and recovery nodes to do their thing as if freshly ticked. 

I don't tentatively think we should also halt the recovery child on recovery action's failure -- but I'm not entirely certain about that

**Comment by SteveMacenski** [comment]
@tonynajjar any update?

**Comment by tonynajjar** [comment]
> Halt the recovery child (1) on main action (0)'s success. Today, nothing is done.

Agree and I think that's what @FrGrQuim  mainly cares about in this issue


> Halt both children (0, 1) on retry. Today, only (0) is halted.

I'm a bit confused because I thought that's what I was proposing too to which you disagreed

> MainAction fails, RecoveryAction succeeds after n attempts, MainAction fails again -> RetryUntilSuccessful should have 5 attempts remaining

Let me clarify this -> MainAction fails, RecoveryAction succeeds after n attempts, MainAction fails again, **RecoveryNode resets child 1 and 2 and goes for try # 2** -> RetryUntilSuccessful should have 5 attempts remaining.

Isn't it equivalent to what you said here?: 
> That obviously changes once we fail MainAction again, because the 2x retries in the RecoveryNode [would reset the first child](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L63-L66). But, shouldn't we actually reset both main action and recovery action children so that a retry gets a clean, new state retry?

**Comment by SteveMacenski** [comment]
So adding a [halt child (1) here](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L53-L56) is OK by all of us, that's settled. Update the Nav2 docs on the RecoveryNode to reflect the update too. 

(summary for myself next time I read this thread 😆 )

---

## Question 1 

> MainAction fails, RecoveryAction succeeds after n attempts, MainAction fails again -> RetryUntilSuccessful should have 5 attempts remaining

Looking back over this, I suppose the `RetryUntilSuccessful` is on the recovery action, which is not retrying until the MainAction is successful, but rather retrying until the recovery is successful. I think this is an odd situation since the SUCCESS/FAILURE is based on if the action succeeded or failed, not necessarily that those recovery actions successfully got the MainAction out of it jam. 

Something like the following would make more sense to me on an application level, retry until the main action is successful and only reset the recount state once it is successfully going again: 

```
<RetryUntilSuccessful num_attempts="5">
  <RecoveryNode number_of_retries="2">
      <MainAction/>
      <RecoveryAction/>
  </RecoveryNode>
</RetryUntilSuccessful>
```

Anyway, I would agree in this example then that after _RecoveryAction succeeds after n attempts_ it should have 5 more attempts. [I think that already happens today though here](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/plugins/control/recovery_node.cpp#L95-L101). When the first child (recovery action) is successful, it is reset. When it fails, it is not. 

> Let me clarify this -> MainAction fails, RecoveryAction succeeds after n attempts, MainAction fails again, RecoveryNode resets child 1 and 2 and goes for try # 2 -> RetryUntilSuccessful should have 5 attempts remaining.

So I think this is already happening.  I don't think any changes are required -- unless you also want it to be reset on failure, which I don't think is the case. 

## Question 2

I said this: 

>   Halt both children (0, 1) on retry. Today, only (0) is halted.
> That would reset the recovery state to fresh once it successfully makes the main action succeed and its done its job. Also, if a retry is made, that retry has a fresh state in both the main and recovery nodes to do their thing as if freshly ticked.

If we start in a clean state, the recovery action (child 2, index 1) should be IDLE so it doesn't need halting. When we trigger it, if its successful it halts itself in Question 1 section above, so its already halted. No need to halt again in the retry. 

If we trigger the recovery action and it fails, we halt the entire system and return `FAILURE` for the parent to this node to handle as it sees fit, no retry is attempted. 

So, I don't think anything needs to change here, right? Even if we did `haltChild(1)` (recovery action), it should always already be halted if we enter that condition?

**Comment by tonynajjar** [comment]
> So I think this is already happening. I don't think any changes are required

From the code you linked I believe it's true yes. By the way, when I mentioned the scenarios I imagined, I didn't base myself off what the node can do today so I didn't mean that something needs to change necessarily, just that this is how I expect the node to behave. Sorry maybe making this clear from the beginning could have reduced some thinking time.

> So, I don't think anything needs to change here, right? (Question 2)

I didn't understand your full train of thought but I am of the same conclusion that nothing else needs to change right now

**Comment by SteveMacenski** [comment]
https://github.com/ros-navigation/navigation2/pull/5027 implements!
